design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
+
+The following list of additional parameters will be supported:
+
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures.
+  * `idp-<identity-provider-alias>` - failing when an IdP is not accessible.
+  * `user-federation-<id>` - failing is LDAP connection is down.
+* (optional) `timeout` - the request gets terminated if not performed within a timeout.
+The implementation needs to take special care of releasing the thread pool (if necessary).
+Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly. If not specified,
+`10s` will be used as default (the same timeout as the one used in Kubernetes).
+
+## Implementation
+
+From the implementation perspective, there are two ways of implementing Healcheck calls within Keycloak:
+* Active - whenever we check if a database is down (this is just an example), we initiate a new connection and
+check if it succeeds.
+* Passive - we position the implementation based on Metrics components (see the Observability Design). We actively
+analyze database connection metrics reported by JCA subsystem
+(see [the currently available metrics in Keycloak](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit#gid=0))
+and if we discover  increasing failure count, we report an at Health Endpoint as well.
+
+The former approach seems more natural but has a lof of drawbacks, including:
+* Additional connections allocated in the thread pools.
+* Often requires integrating with additional subsystems (e.g. JCA for database checks).
+* Complicated to ensure that `timeout` parameter is properly respected.
+* Probing External IdPs might be complicated (e.g. what endpoint to call? what about authentication?)
+
+All the above pain points are not applicable for the Passive Mode. Essentially, all database, User Federation LDAP servers
+and External IdPs probes are free of charge as we report success/failure calls for them in the Metrics components.
+With this approach, the Healthcheck Endpoint becomes an alternative reporting tool for the Metrics.
```

Comment:
Healthcheck endpoints are not designed as a reporting tool for metrics.

They can certainly use metrics to inform their response for liveness/readiness/etc, but it shouldn't be used as a means of reporting metrics that someone is looking at.

Reply:
You're reading this backwards, and I think they're correct here.

If they have already reported/recorded success/failure calls for metrics, using those trends to flag health status makes sense (there is a healthcheck registry in micrometer, e.g., for the same reason: to use calculated sla numbers as a health indicator)

Reply:
I'm sorry, I probably used wrong words here....

The Healthcheck API Design is part of a larger story - the Observability (see [the design](https://github.com/keycloak/keycloak-community/pull/177)). One of the aims of the Observability effort is add more metrics into Keycloak. This Design assumes, that the newly implemented metrics are already implemented. Plus, we have a centralized component, that we can query to get them. 

Having the above in mind, we can query JCA metrics (in case of WF-based deployment) and check the number of failures. If they exceed a certain threshold (to be clarified during the implementation) we could fail the database check. As you can see, both Healthcheck API as well as metrics work in symbiosis in this case. Moreover, the Healthcheck API doesn't do anything actively. It only inspects already collected metrics. So in some sense, from the inner-implementation perspective, it's just another tool that rely on the metrics. 

I hope I made it a bit more clear :)

Reply:
Problem with this approach is that an idle server does not have health checks. Not sure that is a big problem with user federation or external IdPs though. For external user federation we could effectively use the same strategy as for DB, where a connection pool with at least one connection is kept open. For external IdPs that's harder, especially in cases where there are 100s of external IdPs.

Reply:
Let's consider this as two separate cases - DB and External IdPs.

As for the former, the JCA subsystem exposes a couple of [interesting metrics](), including:
- `wildfly_datasources_pool_in_use_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
- `wildfly_datasources_pool_idle_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
- `wildfly_datasources_pool_max_wait_time{data_source="KeycloakDS",microprofile_scope="vendor"}`
- `wildfly_datasources_pool_timed_out{data_source="KeycloakDS",microprofile_scope="vendor"}`
- `wildfly_datasources_pool_created_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
There are many others. I will require some experimentation but I believe we could come up with a pretty good probe only by analyzing the 5 listed above. Another aspect is that JCA subsystem needs to actively probe the datasource. This can be achieved by configuring `check-valid-connection-sql` and `background-validation` (see the [Wildfly Model manual](https://docs.wildfly.org/19/wildscribe/subsystem/datasources/data-source/ExampleDS/index.html)).

External IdPs can be tracked by analyzing the amount of errors in the past few minutes. Perhaps this "few" should be passed into the probe as a parameter or at least there should be some way to override this. If there's no traffic to an external IdP, there will be no errors, thus the probe will pass.

I can think of one, very tricky scenario that I don't have a good answer to. If an External IdP is down for a very long time, Keycloak will flip between ready and not-ready in the predefined intervals. If the Readiness probe is down, Kubernetes will plug Keycloak out of the Load Balancer. Therefore, it will receive no traffic. Since the Readiness probe checks the number of error in the past few minutes, it will eventually report success. Once this happens, Kubernetes will plug Keycloak back into the Load Balancer and Keycloak will receive some traffic. Then, the number of errors reported from an External IdP goes up again and the Readiness probe fails. 

I'm not sure if the above scenario is problematic or not. It's a bit similar to an active Readiness Check where you send some traffic to an External IdP to see if it works again. But in our case, this ends up in a bad user experience. I'm very interested what do you think about this? One solution that comes into my mind is passive/active check. So whenever you're going back from non-ready to ready, you perform an active check (ping the External IdP public URL for example) to see if it's really up.

Reply:
This is a very important question! 

From the user perspective, nothing really changes. Imagine that your deployment requires a Google IdP, which is down for some reason. If you don't include the External IdP check in the Readiness Probe, you will end up with an error screen in Keycloak whenever you try to log in. If you include the External IdP check in the probe, you will get a HTTP 503 Error from Kubernetes/OpenShift as there will be no ready Pods behind a Load Balancer. So the end game is the same - a user can not log in. It's only a matter what Error message he sees.

Also, the same question we should probably ask with User Federation bit. What happens if an LDAP Server is down? Should we ever mark Keycloak as not-ready than?

If we decided to leave both External IdPs and User Federation out of this Design, it will make all the things very simple. The Readiness Check will probably include the database check by default, whereas the Liveness Check will ensure that the REST interface is up. We'd end up with a very simplistic approach that probably won't require any design at all (because it will fit into 2-3 classes).

Reply:
@slaskawi thanks, it does. I was concerned health checks were being treated as more than just utilizing metrics data to determine their status.

Reply:
@slaskawi  @stianst  I would also fall on the side of a readiness probe purpose is to indicate the service is able to serve requests not a particular subset of requests. I don't think you would want to take your keycloak servers out of the service pool just because an IDP was down?  It would still be perfectly valid for a customer to access the server to configure users, new IDPs etc a 503 in this situation would seem the wrong behaviour. I would suggest a focus on whether keycloak can serve requests overall (IE can I access the keycloak server and do regular keycloak admin) If an IDP is unavailable a metric and informational alert may be enough here?

Reply:
Totally agree, @maleck13.

Reply:
I believe this conversation can be resolved now. Together with @stianst we decided to remove External IdPs as well as User Federation bits. See [this comment](https://github.com/keycloak/keycloak-community/pull/55#discussion_r498709614).